### ISSUE TITLE
Wrong Casting Float to String

### ISSUE THREAD

**Issue by arthurdemarchi** [adbailey1/daic_woz_process#3]
Hello, to be honest i'm  not 100% sure this is a problem with the code, or if i've donne something wrong, but it seems that you are sometimes casting the patients folder name (the directories that ranges from 300 to 492) from a float type to a string type to get the paths. But, being treated as float, the cast often results in a path with the floating point  like this: "PATH_TO_PATIENT_FOLDER/300.0/PATIENT_FILE, wich i think was not intended.

Python and types are often confusing since those are not explicitly declared, so as im not sure I'll not make a pull request or anything.

But if you have the time and want to check it out a simple integer cast before the string cast should be enougth to fix this.

If you have the time to answear i'd be glad to know if this code (and the deepAudio Reproduction) are stable, because im facing a lot of 'errors' trying to run them and im not sure if they are in a 'stable' version or if i am missing something.

**Comment by arthurdemarchi** [comment]
Yep, found my mistake, the float was a problem with a edit in the code that I did, due to errors tying to run it,
Anyway, just so the issue is not in vain i'll post the error I was getting prior to my unsuccessful fix that generated the floats in the dataset.

In the merge_csv function, inside utilities, the following line of code:

---comment This checks that the column headers are the same in the two CSV files
    for i in range(len(columnsa)):

should be:


---comment This checks that the column headers are the same in the two CSV files
    for i in range(len(columnsb)):

since columnsa is bigger than columnsb, and you are iterating both.

Im sorry about the first post, I'll close the issue, but i'm still wondering if the code, alongisde you deep audio reproduction repo,  is  stable

**Comment by arthurdemarchi** [comment]
I was able to run the full experiment, here are some changes in code that were needed, I'll keep this here for future reference for those encountering the same problems, And for you, if is your desire to check if they are errors from my wrongfull use of your code or legit bugs.

Again, i'm sorry for the original issue, the wrong type casting was 100% my fault.

audio_file_analysis.py
--#THE PATH FOR THE CONFIG FILES WAS WRONG
		previous:
		    copyfile(workspace_files_dir + '/config.py', current_directory + '/config.py')

		after:
		    copyfile(workspace_files_dir + '/config_files/config.py', current_directory + '/config.py')
utilities.py
--#THE FOR LOOP IN WICH LISTS ARE COMPARED SHOULD ITERATE OVER THE SMALLER LIST
	previous:
	for i in range(len(columnsa)):
		# If headers are different, re-name
		if columnsa[i] != columnsb[i]:
			b = b.rename(columns={columnsb[i]: columnsa[i]})
	
	after:
	for i in range(len(columnsa)):
		# If headers are different, re-name
		if columnsa[i] != columnsb[i]:
			b = b.rename(columns={columnsb[i]: columnsa[i]})


organizer.py
--#THE FULL SPLIT CSV HAS EMPTY LINES THAT NEED TO BE FILTERED
	previous:

	after:
		#Added this condition to data_to_array function
		if file == config.TEST_SPLIT_PATH:
		    data = [x for x in data if x != []]
		    

config_1.py
---#THE DEFAULT TEST CSV POINTED IN CONFIG DO NOT HAVE ENOUGTH COLUMNS FOR THE CODE TO WORK PROPERLY
	previous:
		#TEST_SPLIT_PATH = os.path.join(DATASET, 'test_split_Depression_AVEC2017.csv')
	after:
		TEST_SPLIT_PATH = os.path.join(DATASET, 'full_test_split.csv')
		
(I mean, configs are meant to be changed, but the default configs should work out of the box in my opinion (besides from dataset paths))

**Comment by adbailey1** [comment]
Hi Arthur, 

No problem about your float issue. And thanks for your suggestions here. 

I fixed the copyfile() issue a while ago, but for some reason my push wasn't successful. 

In your utilities.py example, both of your code snippets are the same, could you elaborate on the problem please?

organiser.py - This file is not present in the daic_woz_process, I think it was in the older versions though. 

I can add a note to the readme file for your config.py issue, however, the file names I used for [train_split, dev_split, and test_split] were the file names of the files in the dataset at that the time of downloading them a few years ago. Newer versions of the dataset may have different file names as you suggest. 

Kind regards, 

Andrew

**Comment by arthurdemarchi** [comment]
Hello, Thanks for the answear
The utilities are equal ideed, i'm sorry about that was a typo, but the sugestion I was trying to make was that In the for iterator, instead of iterating until the variable 'columnsa' reaches its end you should be iterating over the variable 'columnsb' since it is smaller, otherwise, when you try to index columnsb for comparing it with columnsa it will be out of bonds

for i in range(len(columnsa)): ---> for i in range(len(columnsb)):

A better correction would be iterating over min(len(columnsb), len(columnsa)) or something like that, a bit more generic. But changing for columnsb was enougth for me

Regarding the original test csv file, at least in the dataset i have acess to, it has only two columns (patient_id, phqbinary) and the function 'data_to_array" assumes all files have at least four columns, so perhaps a check in the file shape before formmating it would be nice and a message explaining what the file was suppose to look like in the readme.

But again, those are just suggestions, i'm already really glad and happy to find you code. Wich is, by the way, really well organised and coded. 

Wonderfull setup on the variables and config I loved using it.

Have a nice week, and thanks for the help1

**Comment by adbailey1** [comment]
Hi Arthur, 

Thank you for the clarifications, I will have a look into them as soon as I can. I am at a conference this week so it will most likely be next week. 

In the meantime, I was wondering whether you would be able to send me the dataset csv files you have over email? I don't want to ask here for dataset privacy reasons. If you are able to, my email is: andrew.bailey@surrey.ac.uk. I can then compare yours against mine and make the changes you recommend to work for both sets.

Also, thanks for your kind words regarding my code, it was much appreciated!

Hope you have a good week too!

Andrew

**Comment by arthurdemarchi** [comment]
Hello Andrew, 

Thank you for taking your time to chat with me, I'll be sending you that e-mail, for patient privacy reasons ill be focused on explain the format of the csv file, the columns names, what data is avaiable and perhaps generate a fake sample or two in the exact format. Instead of simply appending all the data, I hope that's not a problem for you.

I've just redownloaded the dataset last week, so I think (I'm not actually sure) mine is the most updated version, so if you have the time in the future to make the modifications I think it will help a lot of future issues.

although I was able to run the code, i'm stilll facing problems getting the same results as your paper, so I'll work on that for now and try to help if a find anything.

Also, as you've perhaps noticed English  is not my native language so if something gets lost in translation please don't be afraid to point it out.

Thank you again,

Arthur

**Comment by adbailey1** [comment]
Hi Arthur, 

Thanks very much, I am waiting to obtain the dataset myself to check for any differences since I last downloaded but I will use your email as a starting point anyway. 

I assume you are referring to the other repository (https://github.com/adbailey1/DepAudioNet_reproduction), we have noticed variability across devices, so while we were able to reproduce our results on 2 GPU systems, differences in architectures and even running on the CPU may affect performance, which we are currently looking into. 

Your English is very good actually! :) 

Andrew

**Comment by arthurdemarchi** [comment]
Thanks for the response and the commit, i'll update to this version and try to reproduce your results (from that repo) again, i'm also sorry about the late response, i've had some busy days last week.

I'm currently running on a server, i'm not sure wich the gpu system is (nvidia or amd), so i'll look into that. The problem I was facing was a huge overtraining, where acc was 97% on training data and around 40% (basically null) on test data.

If I continue facing this problem i'll try to run on my personal machine (wich is nvdia), and, change some of the settings on the learning rate and optimizer. May take a a bit longer to train, but i'm assuming  it will be closer to your setup.

If you want I can keep you update on that, i'm just not sure where, since technically this is no exactly an issue.

Perhaps I should keep the discussion via e-mail, but that way, it may not help others facing particular same ''problems''.

Thanks again for your time, this was really productive

Until next contact,
Arthur Demarchi

**Comment by adbailey1** [comment]
No problem, I've added a couple more commits since yesterday to fix a few errors I picked up.  

OK that's interesting and thanks for the information. We've noticed some variability across devices so while we can reproduce the results on our servers (and my pc) they may be different on other devices, but shouldn't be dramatically different. 

However, we do not report results on the test set, only the validation set. The test set really shouldn't be used during experimentation as you will begin to overfit the model to the test data (the labels are usually withheld from researchers anyway). 

Consequently, we treat the validation (or development) set as a proxy test set to in an attempt to resolve this problem but by doing so will eventually overfit to this as well. When creating this tool, I focused on trying to replicate the results and approach from DepAudioNet, who I believe used the given training and validation sets. 

However, to resolve these overfitting problems a more standard approach is cross-validation, this combines and splits the training+validation set into multiple subsets. These subsets form our new training set and we hold one out to use as a validation set.  In an early iteration of this tool I incorporated cross-validation, however, it has since been removed but I may work cross-validation back into this tool at some point . 

Hopefully that helps explain some of the decisions in this tool. There are many avenues to explore with this code and dataset as often in research we find ourselves having to balance time spent on prototyping vs software engineering.

Yes please keep me updated on this from your end and my pleasure, happy to help. :-)
